### PR TITLE
Ajax failure triggers, minor tweak to keyboard filter

### PR DESCRIPTION
**Pull request by anutron**
screencap

### CODE REVIEW

File: js/Source/Delegators/Delegator.Ajax.js [hunk 1]
```diff
@@ -84,6 +84,28 @@ name: Delegator.Ajax
         onFailure: function(e){
           if (api.get('errorRedirectURL')){
             window.location.href = api.get('errorRedirectURL');
+          } else if (api.get('failureTriggers')){
+            var delegators = api.getAs(Object, 'failureTriggers');
+            var delegator = api.getDelegator();
+            if (delegators){
+
+              // when the user mouses over/out, loop over the triggers
+              Object.each(delegators, function(config, trigger){
```
Comment (davywentworth) [inline js/Source/Delegators/Delegator.Ajax.js]: your naming is confusing; you loop over `delegators` with each one being called a `trigger`. pick a side! i think the array should probably be called `triggers`.

File: js/Source/Delegators/Delegator.Ajax.js [hunk 2]
```diff
@@ -84,6 +84,28 @@ name: Delegator.Ajax
         onFailure: function(e){
           if (api.get('errorRedirectURL')){
             window.location.href = api.get('errorRedirectURL');
+          } else if (api.get('failureTriggers')){
+            var delegators = api.getAs(Object, 'failureTriggers');
+            var delegator = api.getDelegator();
+            if (delegators){
+
```
Comment (davywentworth) [inline js/Source/Delegators/Delegator.Ajax.js]: remove this empty line.

File: js/Source/Delegators/Delegator.Ajax.js [hunk 3]
```diff
@@ -84,6 +84,28 @@ name: Delegator.Ajax
...
+              // when the user mouses over/out, loop over the triggers
+              Object.each(delegators, function(config, trigger){
+                // split the trigger name - '.foo::addClass' > {name: addClass, selector: .foo}
+                trigger = delegator._splitTriggerName(trigger);
+                if (!trigger) return;
+
```
Comment (davywentworth) [inline js/Source/Delegators/Delegator.Ajax.js]: remove this empty line.

File: sandbox/docs/JavaScript/Delegators/Delegator.Ajax.html.erb [hunk 4]
```diff
@@ -102,6 +102,55 @@
...
+
+    <pre><code>&lt;a class="btn btn-default"
+    data-trigger="ajax Stop"
+    data-ajax-options="
+      {
+        'useSpinner':true,
+        'action':'replace',
+        'target':'self',
+        'failureTriggers':{
+          '!&gt;div::addClass':{
```
Comment (davywentworth) [inline sandbox/docs/JavaScript/Delegators/Delegator.Ajax.html.erb]: i feel like the syntax would be much easier to read if it was
```
failureTriggers': {
  'addClass': {
    'triggerSelector': '>div',
    'class': 'red'
  }
}

File: js/Source/Delegators/Delegator.Ajax.js [hunk 5]
```diff
@@ -84,6 +84,28 @@ name: Delegator.Ajax
...
+                // iterate over the elements that match that selector using the event target as the root
+                Behavior.getTargets(link, trigger.selector).each(function(target){
+                  var api;
+                  // create an api for the trigger/element combo and set defaults to the config (if config present)
+                  if (config) api = delegator._getAPI(target, trigger).setDefault(config);
+                  // invoke the trigger
+                  delegator.trigger(trigger.name, target, event, true, api);
+                });
+              });
+
```
Comment (davywentworth) [inline js/Source/Delegators/Delegator.Ajax.js]: remove this empty line

File: js/Source/Delegators/Delegator.Ajax.js (hunk 1 above)
Comment (anutron) [inline js/Source/Delegators/Delegator.Ajax.js]: 👍

File: sandbox/docs/JavaScript/Delegators/Delegator.Ajax.html.erb (hunk 4 above)
Comment (anutron) [inline sandbox/docs/JavaScript/Delegators/Delegator.Ajax.html.erb]: it's a convention that already exists in other things. `selector::trigger: {config}`

Comment (anutron) [inline sandbox/docs/JavaScript/Delegators/Delegator.Ajax.html.erb]: Not necessarily disagreeing with you. Just saying there's already prior work in place that we'd have to deprecate.

**Pull request closed without merging**